### PR TITLE
Add for-each loops for iterating over map elements

### PR DESCRIPTION
**Pull request by ajor**
Basic support for #2955. Support for sharing variables will be added in a separate PR: #3014

For loops are implemented using the bpf_for_each_map_elem helper function, which requires them to be rewritten into a callback style.

In this first implementation, we do not allow any variables to be shared between the main probe and the loop's body. Maps are global so are not a problem.

Pseudo code for the transformation we apply:

Before:
```
PROBE {
  @map[0] = 1;
  for ($kv : @map) {
    [LOOP BODY]
  }
}
```

After:
```
PROBE {
  @map[0] = 1;
  bpf_for_each_map_elem(@map, &map_for_each_cb, 0, 0);
}
long map_for_each_cb(bpf_map *map, const void *key, void *value, void *ctx) {
  $kv = ((uint64)key, (uint64)value);
  [LOOP BODY]
}
```

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

File: src/ast/passes/printer.cpp
Comment (jordalgo) [inline src/ast/passes/printer.cpp:15]: Just curious, why increment the initial depth before visiting the root?

Comment (jordalgo) [reply src/ast/passes/printer.cpp:15]: NM, I see what you're doing.

File: src/ast/passes/semantic_analyser.cpp
Comment (jordalgo) [inline src/ast/passes/semantic_analyser.cpp:1924]: You mention the kernel version in the docs, want to also mention it here?

File: tests/semantic_analyser.cpp
Comment (jordalgo) [inline tests/semantic_analyser.cpp:3017]: This indentation looks off.

File: tests/semantic_analyser.cpp
Comment (jordalgo) [inline tests/semantic_analyser.cpp:3057]: Should we add the negative test to show it errors with the wrong type.

File: tests/semantic_analyser.cpp
Comment (jordalgo) [inline tests/semantic_analyser.cpp:3110]: nit
```suggestion
stdin:5:9-19: ERROR: Variables defined outside of a for-loop can not be accessed in the loop's scope
```

File: src/ast/passes/semantic_analyser.cpp
Comment (jordalgo) [inline src/ast/passes/semantic_analyser.cpp:1983]: nit
```suggestion
                                   "can not be accessed in the loop's scope";
```

File: src/ast/passes/semantic_analyser.cpp
Comment (jordalgo) [inline src/ast/passes/semantic_analyser.cpp:1995]:
```suggestion
        << "'" << opstr(n) << "' statement is not allowed in a for-loop";
```

File: src/ast/passes/semantic_analyser.cpp
Comment (jordalgo) [inline src/ast/passes/semantic_analyser.cpp:924]: JC, what is this additional check for?

File: tests/runtime/for
Comment (jordalgo) [reply tests/runtime/for:29]: Can we also add a test for two maps and two loops?

File: man/adoc/bpftrace.adoc
Comment (jordalgo) [inline man/adoc/bpftrace.adoc:872]: I think also showing an example with multiple keys would be good too.

**Comment by jordalgo** [comment]
Overall this is looking good! Few minor nits/questions.

File: tests/semantic_analyser.cpp
Comment (ajor) [reply tests/semantic_analyser.cpp:3017]: In the AST, the value being assigned to the map is a child of the AssignMapStatement node (shown as "=" by Printer).

This is the tree that we're trying to show:

```
     AssignMapStatement (=)
      /                \
     / LHS              \ RHS
    /                    \
  Map (@map)          Integer (1)
   |
   | key
   |
Integer (0)
```

File: src/ast/passes/semantic_analyser.cpp
Comment (ajor) [reply src/ast/passes/semantic_analyser.cpp:1924]: Our feature detection logic means that this can run on older kernels if they've backported the required helper, so mentioning it in the error message could be misleading.

I wasn't sure about including it in the docs, but put it in because the docs for while loops mentioned it too. It's nice to have a rough idea of when a feature is available, but backporting happens in a few distros so having it in the docs could be misleading too... 🤷‍♂️

File: src/ast/passes/semantic_analyser.cpp
Comment (ajor) [reply src/ast/passes/semantic_analyser.cpp:2014]: It can happen before type-propagation is complete if a map is defined after being used in a for-loop, e.g.:

`BEGIN { for ($kv : @map) {} } END { @map[0] = 1; }`

`@map` doesn't exist on the first semantic analyser pass.

File: src/ast/passes/semantic_analyser.cpp
Comment (ajor) [reply src/ast/passes/semantic_analyser.cpp:924]: `print()` comes in two forms: print an entire map or print a single value.

Only the first form is asynchronous, so the second is safe to use inside loops. The error message below was already assuming users were printing an entire map (1st form), so really this check had been wrong for a long time, but I just noticed it when I was testing for-loops.

File: src/ast/visitors.cpp
Comment (ajor) [reply src/ast/visitors.cpp:160]: Yeah that sounds like a reasonable optimisation to do. I think right now it'd have to be done in CodeGen. I'll leave it out of this PR as it's already complicated enough though.

This is the kind of thing where it'd be nice to have an AST rewriter: we could have a really simple pass which just deletes empty for-loops and CodeGen wouldn't need any extra logic.

File: tests/runtime/for
Comment (ajor) [reply tests/runtime/for:29]: JSON output I'd say is orthogonal to for-loops. As long a text output works with for-loops, and JSON output works without for-loops, then JSON output will work with for-loops, so no need to test the combination.

Comment (ajor) [reply tests/runtime/for:29]: Testing two loops: good idea! I discovered a bug using nested loops just now:
`BEGIN { @map1[0] = 1; @map2[0] = 1; for ($kv1 : @map1) { for ($kv2 : @map2) {} } }`

This triggers a verifier error: "call to invalid destination". Will need to look into what that means.

File: tests/semantic_analyser.cpp
Comment (ajor) [reply tests/semantic_analyser.cpp:3057]: Like JSON output, I think variable assignments are generally orthogonal to for-loops. I can't actually remember why I added this test in the first place. Will remove it if I can't work it out.

I'd think that all we need to check is that `$kv` has the correct type, and the test case `for_loop_map_one_key` is already doing that by printing out the types in the AST.

**Comment by ajor** [comment]
Addressed most of the comments. Still need to add tests for multiple for-loops and fix the nested for-loop bug.

File: src/ast/passes/semantic_analyser.cpp
Comment (jordalgo) [reply src/ast/passes/semantic_analyser.cpp:1924]: I think it's ok to be verbose in these scenarios e.g. mention the kernel version but it may be specific to the distro if the changes were backported.

File: tests/semantic_analyser.cpp
Comment (jordalgo) [reply tests/semantic_analyser.cpp:3017]: Ah yeah, my bad, I wasn't noticing the equal assignment being the parent.

File: src/ast/visitors.cpp
Comment (jordalgo) [reply src/ast/visitors.cpp:160]: > This is the kind of thing where it'd be nice to have an AST rewriter: we could have a really simple pass which just deletes empty for-loops and CodeGen wouldn't need any extra logic.

Yah, that would be cool. I can maybe take a look at doing that in a follow-up (just for my own education).

File: src/ast/passes/semantic_analyser.cpp
Comment (jordalgo) [reply src/ast/passes/semantic_analyser.cpp:924]: Ah gotcha. Thanks for the explanation.

**Comment by ajor** [comment]
Added tests for multiple loops in various scenarios.

Nested loops requires another relocation fix: #3020
Loops used from two different main programs is causing yet another error! 
> number of funcs in func_info doesn't match number of subprogs

This one appears to have been around for longer and affects the len() builtin as well:
`BEGIN { @map[0] = 1; print(len(@map)) } END { print(len(@map)) }`

**Comment by ajor** [comment]
The loops-in-multiple-probes issue is pretty nasty: #3021

File: src/ast/passes/collect_nodes.h
Comment (jordalgo) [inline src/ast/passes/collect_nodes.h:28]: nit: I prefer `get_nodes`

**Review by jordalgo — APPROVED** [review]
This looks great. Nice work! 🚀

**Comment by danobi** [comment]
I would like to take a look but I'll be busy until ~Tuesday.

File: src/ast/passes/semantic_analyser.cpp
Comment (viktormalik) [inline src/ast/passes/semantic_analyser.cpp:2000]: Nit: you could do the same as semantic analyser does at the end:
```suggestion
  if (!err_.empty())
    return;
```
That would spare the `error` variable.

Maybe we could even introduce some wrapper like `has_error` as it may be useful in other places, too.

**Review by viktormalik — COMMENTED** [review]
Just a couple of small comments, otherwise looks great!

**Comment by viktormalik** [comment]
Also codegen tests need update after #2985 was merged.

File: src/ast/passes/semantic_analyser.cpp
Comment (danobi) [inline src/ast/passes/semantic_analyser.cpp:1976]: Neat :)

**Review by danobi — COMMENTED** [review]
overall looks pretty reasonable. didn't get a chance to look carefully let

File: man/adoc/bpftrace.adoc
Comment (ajor) [reply man/adoc/bpftrace.adoc:549]: There's some discussion on syntax in #2963

File: src/ast/passes/collect_nodes.h
Comment (ajor) [reply src/ast/passes/collect_nodes.h]: I've got to disagree on this one :)

**Comment by ajor** [comment]
Addressed review comments and added extra runtime test for deleting elements while iterating over the map.

**Comment by ajor** [comment]
Fixed a lifetime bug I'd introduced for tuple elements: in `CodegenLLVM::visit(Tuple &tuple)` I'm now extending the lifetime of all tuple elements until the end of the function with:
```
scoped_dels.emplace_back(std::move(accept(elem)));
```

Thanks @arthurshau for the new test case which exposed this! (`call.print_map_item_tuple`).

Will merge tomorrow if there are no objections.

**Review by viktormalik — APPROVED** [review]

**Comment by ajor** [comment]
A lot of the error messages I'm testing for in this PR appear to have regressed: #3063

**Comment by ajor** [comment]
Updated the tests to accept the incorrect expected results and left comments referring to #3063

**Pull request merged**
